// @reactivex/rxjs/dist/es6/operator/finally.js
import { Subscriber } from '../Subscriber';
import { Subscription } from '../Subscription';
/**
 * Returns an Observable that mirrors the source Observable, but will call a specified function when
 * the source terminates on complete or error.
 * @param {function} finallySelector function to be called when source terminates.
 * @return {Observable} an Observable that mirrors the source, but will call the specified function on termination.
 * @method finally
 * @owner Observable
 */
export function _finally(finallySelector) {
    return this.lift(new FinallyOperator(finallySelector));
}
class FinallyOperator {
    constructor(finallySelector) {
        this.finallySelector = finallySelector;
    }
    call(subscriber, source) {
        return source._subscribe(new FinallySubscriber(subscriber, this.finallySelector));
    }
}
/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends {Ignored}
 */
class FinallySubscriber extends Subscriber {
    constructor(destination, finallySelector) {
        super(destination);
        this.add(new Subscription(finallySelector));
    }
}
